pkg/logging/logging.go

```diff
@@ -0,0 +1,256 @@
+type VerbosityLevel int
+
+const (
+	MANDATORY VerbosityLevel = 0
+	DEBUG     VerbosityLevel = 1
+	TRACE     VerbosityLevel = 2
+)
+
+// LogEntry defines the information for a single entry stored in the system logs.
+// It is assumed that the timestamp is added automatically and thus not included in the log entry structure.
+// Verbosity and log type are added by the log library via which the entries are sent to stdout or stderr.
+type LogEntry struct {
+	// Caller indicates the cluster, component and function that generated the message in the format "cluster/component/function"
+	// Use the buildCallerPath function to construct this.
+	// +required
+	Caller string `json:"caller"`
+
+	// Message being logged.
+	// +required
+	Message string `json:"message"`
+
+	// Unique id of the kubernetes fybrikapplication (Needs to be unique over time in order to support history and not just current workloads)
+	// +required
+	FybrikAppGUID string `json:"fybrikAppGuid"`
+
+	// ForUser should be true if the message should be shared with the end user in the FybrikApplication status.
+	// Assumed false if not indicated.
+	// +optional
+	ForUser bool `json:"forUser,omitempty"`
+
+	// ForArchive should be true if the message should be archived long term.
+	// For example, contains full contents of FybrikApplication and its status and should be stored for auditing purposes.
+	// Assumed false if not indicated
+	// +optional
+	ForArchive bool `json:"forArchive,omitempty"`
+
+	// DataSetID is a unique identifier for the data set.
+	// It should include an indicator of the catalog from which it came, as well as the catalog id.
+	// +optional
+	DataSetID string `json:"datasetID,omitempty"`
+
+	// ResponseTime of the current operation in milliseconds.
+	// This is an optional parameter that can use in monitoring dashboards such as Kibana.
+	// +optional
+	ResponseTime int64 `json:"responseTime,omitempty"`
+
+	// Action is the current operation being called.
+	// For example, “create_plotter” or “update_blueprint”.
+	// +optional
+	Action string `json:"action,omitempty"`
+}
+
+// ------ User Facing Log Functions ---------
+// Components (control plane and data plane) that interface with the user / user workload should log interactions their USER LEVEL log entries via user level logging functions. LogFAError, LogFAStruct, LogFAUpdate
+// This ensures that they are logged with ForUser and ForArchive set to true
+// Log entries not relevant to users should be logged with the internal logging functions
+
+// LogFAStruct writes to the log file a log entry containing for example, the FybrikApplication structure or a subset (spec or status) of it.
+// The log entry indicates it's a user level entry and should be archived.
+// This could be used by data path components as well to log user level structures.
+// It is logged as INFO/MANDATORY
+func LogFAStruct(log logr.Logger, arg interface{}, argName string, caller string, guid string, cluster string, function string) {
```

Comment:
@simanadler In this and next functions `log` is used. Do you intend to implement Logger interface? Standard Logger implementations, including the one currently used, are not printing info as JSON. 
What Logger implementation will be used for the main manager process? My concern is that we will have a mixed structure in logs that will be not identifiable by external tools.

Reply:
@shlomitk1 Based on the comments to this PR and my discussion with @roytman this morning, I'm looking at moving to https://github.com/rs/zerolog instead of go-logr, since it supports json based structured logging.